sdks/java/core/src/main/java/org/apache/beam/sdk/io/UnboundedReadFromBoundedSource.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Override
+    public void validate() {
+      boundedSource.validate();
+    }
+
+    @Override
+    public List<BoundedToUnboundedSourceAdapter<T>> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+      long desiredBundleSize = boundedSource.getEstimatedSizeBytes(options) / desiredNumSplits;
+      List<BoundedToUnboundedSourceAdapter<T>> result = Lists.newArrayList();
+      for (BoundedSource<T> split : boundedSource.splitIntoBundles(desiredBundleSize, options)) {
+        result.add(new BoundedToUnboundedSourceAdapter<>(split));
+      }
+      return result;
+    }
+
+    @Override
+    public Reader createReader(PipelineOptions options, Checkpoint checkpoint) throws IOException {
+      return new Reader(boundedSource.createReader(options), checkpoint);
+    }
+
+    @Override
+    public Coder<T> getDefaultOutputCoder() {
+      return boundedSource.getDefaultOutputCoder();
+    }
+
+    static class Checkpoint implements UnboundedSource.CheckpointMark {
+      private final boolean done;
+
+      public Checkpoint(boolean done) {
+        this.done = done;
+      }
+
+      public boolean isDone() {
+        return done;
+      }
+
+      @Override
+      public void finalizeCheckpoint() {}
+    }
+
+    @Override
+    public Coder<Checkpoint> getCheckpointMarkCoder() {
+      return AvroCoder.of(Checkpoint.class);
+    }
+
+    private class Reader extends UnboundedReader<T> {
+      private final BoundedSource.BoundedReader<T> boundedReader;
+      private boolean done;
+
+      public Reader(BoundedSource.BoundedReader<T> boundedReader, Checkpoint checkpoint) {
+        this.done = checkpoint != null && checkpoint.isDone();
+        this.boundedReader = boundedReader;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        if (done) {
+          return false;
+        }
+        boolean result = boundedReader.start();
+        if (!result) {
+          done = true;
+          boundedReader.close();
+        }
+        return result;
+      }
+
+      @Override
+      public boolean advance() throws IOException {
+        if (done) {
+          return false;
+        }
+        boolean result = boundedReader.advance();
+        if (!result) {
+          done = true;
+          boundedReader.close();
+        }
+        return result;
+      }
+
+      @Override
+      public void close() {}
+
+      @Override
+      public T getCurrent() throws NoSuchElementException {
+        return boundedReader.getCurrent();
+      }
+
+      @Override
+      public Instant getCurrentTimestamp() throws NoSuchElementException {
+        return boundedReader.getCurrentTimestamp();
+      }
+
+      @Override
+      public Instant getWatermark() {
+        return done ? BoundedWindow.TIMESTAMP_MAX_VALUE : BoundedWindow.TIMESTAMP_MIN_VALUE;
+      }
+
+      @Override
+      public Checkpoint getCheckpointMark() {
+        return new Checkpoint(done);
```

Comment:
This does not retain state appropriately - if the runner gets a checkpoint in the middle of reading a source, you will restart from the beginning of the source when provided the checkpoint.